crates/env/src/engine/mod.rs

```diff
@@ -25,7 +25,7 @@ pub trait OnInstance: EnvBackend + TypedEnvBackend {
 }
 
 cfg_if! {
-    if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
+    if #[cfg(any(all(not(feature = "std"), target_arch = "wasm32"), feature = "ink-experimental-engine"))] {
```

Comment:
The `ink_env` crate is meant to do this dispatch between on-chain and off-chain environments. So please also remove this logic from any particular environment implementation.